Added samples per second and remaining time information to progress bar

jt-lab opened the pull request:
I found myself quite often copying the "samples complete" and "time elapsed" info out of the terminal to calculate the sampling speed (to compare different step methods or computers) or the remaining time. (often accidentally aborting sampling by pressing ctrl+C...).

Therefore, I added SPS (samples per second) and ETE (estimated time enroute) to the progressbar... I find it quite helpful and would suggest something like this to be added (maybe optionally).

Commit:
Added samples per second and remaining time information to progress bar

springcoil commented:
I think this is a fair thing to add. What do you think @twiecki?

On 22 Aug 2016 11:26 PM, "jeti182" notifications@github.com wrote:

> I found myself quite often copying the "samples complete" and "time
> elapsed" info out of the terminal to calculate the sampling speed (to
> compare different step methods or computers) or the remaining time. (often
> accidentally aborting sampling by pressing ctrl+C...).
> 
> Therefore, I added SPS (samples per second) and ETE (estimated time
> enroute) to the progressbar... I find it quite helpful and would suggest
> 
> ## something like this to be added (maybe optionally).
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/pymc-devs/pymc3/pull/1296
> Commit Summary
> - Added samples per second and remaining time information to progress
>   bar
> 
> File Changes
> - _M_ pymc3/progressbar.py
>   https://github.com/pymc-devs/pymc3/pull/1296/files#diff-0 (10)
> 
> Patch Links:
> - https://github.com/pymc-devs/pymc3/pull/1296.patch
> - https://github.com/pymc-devs/pymc3/pull/1296.diff
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/1296, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AA8DiNJ30Yyky9Y546DbFKe7aLXDZVnaks5qiiIQgaJpZM4JqXsM
> .

twiecki commented:
Yeah, this looks great. My only concern is that people won't know what the abbreviations stand for, specifically ETE. Could we instead use ETA?

jt-lab commented:
I thought about this, too. Strictly, ETA would be wrong.On the other hand it is often used even in such a context (https://de.wikipedia.org/wiki/Estimated_time_of_arrival; sorry German text, but the engl. Wikipedia does not contain the discussion or ETA vs ETE). Another suggestion is TTG (time to go)  which I find more weird, however. 

From the formatting of the number and their stready decrease I think it is pretty obvious what this is. The SPS may need explanation, however. Maybe when sampling starts there could be a little info text?

twiecki commented:
ETA is pretty common so I think that's a good reason to use it here (even if it's not completely correct). SPS might be fine too and could be described in doc string of the sampling function.

jt-lab commented:
OK.

If you want I can implement and test this as well for ipython notebook progress bar

twiecki commented:
@jeti182 that'd be fantastic!

jt-lab commented:
Hmm, is it possible that the IPython / Jupyter integration is broken? I'm having trouble running pymc3 sampling with the ipython progress widget. The default progressbar works fine. 

Apparently, in interactive_sampling.py the assertion:

`assert(hasattr(IPython.get_ipython(), 'comm_manager'))`

fails.

I was trying now with jupyter, but had the same (or at least very similar problem with iypthon).

Is anyone successfully using the widgets in notebooks? 
If not, maybe the widget integration should be removed all together. I saw some trouble with that in earlier issues. And since I've never seen it working I'm not sure what the benefit over the usual progress indicator is (beyond fanciness). But I'd love to see it ... any ideas?

twiecki commented:
Maybe we leave it then for another PR but get the ETA stuff in?

twiecki commented:
@jeti182 any progress? I think  once we renamed to ETA and added this to docs we can merge.

Commit:
- Changed ETE to ETA in progressbar.py
- Removed unnecessary casts progressbar.py
- Reduced witdh of progress bar in progressbar.py

- Added explanations for SPS and ETA in the docstring of
  "sample" in sampling.py
- Added missing period in the docstring for another parameter

Commit:
- Fixed typo

jt-lab commented:
I just pushed these changes:
1.  changed ETE to ETA
2. added the explanation of the abbreviations in the docstring of sample in sampling for the progress bar flag. Formerly it was:

"Flag for progress bar"

I now changed it to:

 "     Whether or not to display a progress bar in the command line. The 
        bar shows the percentage of completion, the sampling speed in 
        samples per second (SPS), and the estimated remaining time until
        completion ("expected time of arrival"; ETA).

"
3. some minor stuff (see commit log)

twiecki commented on pymc3/progressbar.py:
```diff
@@ -54,7 +55,12 @@ def animate(self, i, elapsed):
-        return "[%s] %i of %i complete in %.1f sec" % (bar, i, self.iterations, round(elapsed, 1))
+        sps = i / float(elapsed) 
+        eta = (self.iterations / sps) - elapsed
+        return ("[%s] %i of %i in %.1f sec | SPS: %.1f | ETA: %s" % 
```
I think a `.format()` would be nicer here:
`"[{bar}] {iteration} of {total_iterations} in ...".format(bar=bar, ...)`

Commit:
Use fromat instead of %.. place holer fromat string.

jt-lab commented:
Using format leads to a terribly long line ... However, I think I came up with an acceptable solution ... see commit ...

twiecki commented on pymc3/progressbar.py:
```diff
@@ -54,7 +55,18 @@ def animate(self, i, elapsed):
+
+        prog_str = "[{bar}] {it} of {its} in {s_elapsed} sec. " \
+                   "| SPS: {sps} | ETA: {eta}" \
+                   "".format(bar = bar, it = i, its = its, 
```
PEP8 suggests no spaces around kwargs: `.format(bar=bar, it=i...)`

jt-lab committed:
Update progressbar.py

Removed spaces around "=" in keyword arguments

Commit:
autopep8ed file

twiecki commented:
:+1: thanks!
